packages/java-edition/src/binder/index.ts

```diff
@@ -10,130 +9,145 @@ import type {
 import { fileUtil, TaggableResourceLocationCategories } from '@spyglassmc/core'
 import { ReleaseVersion } from '../dependency/index.js'
 
+interface ScopedVersion {
+	since?: ReleaseVersion
+	until?: ReleaseVersion
+}
+
+interface CategoryDef extends ScopedVersion {
+	category: FileCategory | TagFileCategory
+	extname: string
+	pack: 'data_pack'
+}
+function resource(
+	pack: 'data_pack',
+	category: FileCategory | TagFileCategory,
+	path: string | ScopedVersion,
+	version?: ScopedVersion,
+): [string, CategoryDef] {
+	if (typeof path !== 'string') {
+		// Shift parameters if path is a scoped version
+		version = path
+		path = category
+	}
+	return [path, { pack, category, extname: '.json', ...version }]
```

Comment:
instead of relying on positional-argument jank, can we just make this an object so the parameter position doesn't matter?

```ts
function resource(
	pack: 'data_pack',
	{ category, path, version }: {
		category: FileCategory | TagFileCategory,
		path: string,		
		version?: ScopedVersion,
	}
): [string, CategoryDef] {
	return [path, { pack, category, extname: '.json', ...version }]
```

Reply:
I think the reason for the positional arguments (I wouldn't say it's jank), it to keep the function calls short in the map below.

Reply:
ahh that's fair